### PR TITLE
apps: code cleanup, initial typing, debug functionality

### PR DESCRIPTION
**Pull request by martinpitt**
This helped me with https://bugzilla.redhat.com/show_bug.cgi?id=2325759

### CODE REVIEW

File: pkg/apps/utils.tsx [hunk 1]
```diff
@@ -27,14 +27,27 @@ import { show_modal_dialog } from "cockpit-components-dialog.jsx";
 
 const _ = cockpit.gettext;
 
-export function icon_url(path_or_url) {
+export function debug(...args: unknown[]) {
+    if (window.debugging == "all" || window.debugging?.includes("apps")) {
+        console.debug("apps:", ...args);
```
Comment (cockpituous) [inline pkg/apps/utils.tsx:32]: These 2 added lines are not executed by any test.

File: pkg/apps/utils.tsx [hunk 2]
```diff
@@ -89,9 +110,29 @@ export const show_error = ex => {
...
+    summary: string,
+    description: string,
+    icon: string, // this is a path
+    screenshots: { full: string }[],
+    launchables: Launchable[],
+    urls: { type: string, link: string }[],
+    installed?: boolean,
+};
+
+export const launch = (comp: Component) => {
```
Comment (cockpituous) [inline pkg/apps/utils.tsx:132]: This added line is not executed by any test.

File: pkg/apps/utils.tsx [hunk 3]
```diff
@@ -89,9 +110,29 @@ export const show_error = ex => {
...
     for (let i = 0; i < comp.launchables.length; i++) {
         if (comp.launchables[i].type == "cockpit-manifest") {
+            debug("launching", comp.launchables[i].name, "in component", JSON.stringify(comp));
```
Comment (cockpituous) [inline pkg/apps/utils.tsx:135]: This added line is not executed by any test.

File: pkg/apps/utils.tsx [hunk 4]
```diff
@@ -27,14 +27,21 @@ import { show_modal_dialog } from "cockpit-components-dialog.jsx";
...
-export function icon_url(path_or_url) {
+export function icon_url(path_or_url: string): string {
     if (!path_or_url)
         return "default.png";
 
     if (path_or_url[0] != '/')
         return path_or_url;
 
-    const queryobj = {
+    type QueryObj = {
```
Comment (mvollmer) [inline pkg/apps/utils.tsx:43]: I think `interface` is preferred over `type`. Can't say why, however.

File: pkg/apps/utils.tsx [hunk 5]
```diff
@@ -49,7 +56,11 @@ export function icon_url(path_or_url) {
     return prefix + '?' + query;
 }
 
-export const ProgressBar = ({ size, title, data, ariaLabelledBy }) => {
+export const ProgressBar = ({ size, data, ariaLabelledBy }: {
+    size?: "lg" | "md" | "sm",
+    data: { percentage: number, waiting: boolean },
```
Comment (mvollmer) [inline pkg/apps/utils.tsx:67]: I guess `data` is actually the same object that is also passed to `CancelButton`. Could we make a real type (or interface) for it?

Comment (mvollmer) [reply pkg/apps/utils.tsx:67]: `interface JobData { percentage: number, waiting: boolean, cancel: ... }`. I haven't checked the callers yet....

File: pkg/apps/utils.tsx (hunk 4 above)
Comment (martinpitt) [reply pkg/apps/utils.tsx:43]: Oh, ok -- in my mental model `type` is  similar to `struct` (i.e. data only), and `interface` similar to `class` (i.e. behaviour/methods). We use `type` a lot for shaping data, e.g. in pkg/lib/cockpit/channel.ts, but that also has e.g. `interface BaseChannelOptions` which is also data only The `interface`s in pkg/lib/cockpit.d.ts are all class-y (with methods).

https://blog.logrocket.com/types-vs-interfaces-typescript/ shows some difference, like unions (type only) or merging and "subclassing" with `extends` (interface only).

So, I don't have a strong opinion on that, and right now we don't have much consistency here. I'm fine with moving to `interface`.

Comment (mvollmer) [reply pkg/apps/utils.tsx:43]: Yeah, thanks for investigating. I had this paragraph from the TS Handbook in mind:

For the most part, you can choose based on personal preference, and TypeScript will tell you if it needs something to be the other kind of declaration. If you would like a heuristic, use interface until you need to use features from type.

https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#interfaces

I guess we can always change a type to an interface without affecting the users, so... *shrug* :-)

Comment (martinpitt) [reply pkg/apps/utils.tsx:43]: > If you would like a heuristic, use interface until you need to use features from type.

That's a good piece of advice, thanks!

File: pkg/apps/application.jsx [hunk 6]
```diff
@@ -104,7 +103,7 @@ export const Application = ({ metainfo_db, id, progress, progress_title, action
 
         let progress_or_launch;
         if (progress) {
-            progress_or_launch = <ProgressBar title={progress_title} data={progress} />;
+            progress_or_launch = <ProgressBar data={progress} />;
```
Comment (cockpituous) [inline pkg/apps/application.jsx:106]: This added line is not executed by any test.

File: pkg/apps/utils.tsx (hunk 1 above)
Comment (cockpituous) [inline pkg/apps/utils.tsx:32]: These 2 added lines are not executed by any test.

File: pkg/apps/utils.tsx [hunk 7]
```diff
@@ -64,12 +87,14 @@ export const ProgressBar = ({ size, title, data, ariaLabelledBy }) => {
...
-export const CancelButton = ({ data }) => (
+export const CancelButton = ({ data }: { data: JobProgress }) => (
     <Button variant="secondary" isDisabled={!data.cancel} onClick={data.cancel}>
         {_("Cancel")}
     </Button>);
 
-export const show_error = ex => {
+// ex is a PackageKit error; requires typing pkg/lib/packagekit.js first
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const show_error = (ex: any) => {
```
Comment (cockpituous) [inline pkg/apps/utils.tsx:97]: This added line is not executed by any test.

File: pkg/apps/utils.tsx [hunk 8]
```diff
@@ -89,9 +114,29 @@ export const show_error = ex => {
...
+    summary: string,
+    description: string,
+    icon: string, // this is a path
+    screenshots: { full: string }[],
+    launchables: Launchable[],
+    urls: { type: string, link: string }[],
+    installed?: boolean,
+}
+
+export const launch = (comp: Component) => {
```
Comment (cockpituous) [inline pkg/apps/utils.tsx:136]: This added line is not executed by any test.

File: pkg/apps/utils.tsx [hunk 9]
```diff
@@ -89,9 +114,29 @@ export const show_error = ex => {
...
     for (let i = 0; i < comp.launchables.length; i++) {
         if (comp.launchables[i].type == "cockpit-manifest") {
+            debug("launching", comp.launchables[i].name, "in component", JSON.stringify(comp));
```
Comment (cockpituous) [inline pkg/apps/utils.tsx:139]: This added line is not executed by any test.

**Review by mvollmer — APPROVED** [review]
Thanks!

**Pull request merged**